File: SQL/Expressions/SQLBitwiseExpression.vb

' ___________________________________________________
'
'  © Hi-Integrity Systems 2012. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On


Namespace SQL

    Public Enum BitwiseOperator
        [And]
        [Or]
    End Enum

    Public Class SQLBitwiseExpression
        Inherits SQLExpression

        Private pobjLeft As SQLExpression
        Private peOperator As BitwiseOperator
        Private pobjRight As SQLExpression

        Public Sub New()

        End Sub

        Public Sub New(ByVal strLeftFieldName As String, ByVal eOperator As BitwiseOperator, ByVal objRightValue As Object)

            Me.New(New SQLFieldExpression(strLeftFieldName), eOperator, New SQLValueExpression(objRightValue))

        End Sub

        Public Sub New(ByVal strLeftFieldName As String, ByVal eOperator As BitwiseOperator, ByVal objRightExpression As SQLExpression)

            Me.New(New SQLFieldExpression(strLeftFieldName), eOperator, objRightExpression)

        End Sub

        Public Sub New(ByVal objLeftExpression As SQLExpression, ByVal eOperator As BitwiseOperator, ByVal objRightExpression As SQLExpression)

            pobjLeft = objLeftExpression
            peOperator = eOperator
            pobjRight = objRightExpression

        End Sub

        Public Property LeftExpression() As SQLExpression
            Get

                Return pobjLeft

            End Get

            Set(ByVal value As SQLExpression)

                If value Is Nothing Then
                    Throw New ArgumentNullException
                End If

                pobjLeft = value

            End Set
        End Property

        Public Property RightExpression() As SQLExpression
            Get

                Return pobjRight

            End Get

            Set(ByVal value As SQLExpression)

                If value Is Nothing Then
                    Throw New ArgumentNullException
                End If

                pobjRight = value

            End Set
        End Property

        Public Property [Operator]() As BitwiseOperator
            Get

                Return peOperator

            End Get

            Set(ByVal value As BitwiseOperator)

                peOperator = value

            End Set
        End Property

        Friend Overrides Function SQL(ByVal eConnectionType As Database.ConnectionType) As String

            If pobjLeft Is Nothing Then
                Throw New ArgumentNullException(Me.GetType.Name & ".LeftExpression")
            ElseIf pobjRight Is Nothing Then
                Throw New ArgumentNullException(Me.GetType.Name & ".RightExpression")
            End If

            Return "(" & pobjLeft.SQL(eConnectionType) & " " & OperatorString(peOperator) & " " & pobjRight.SQL(eConnectionType) & ")"

        End Function

        Private Function OperatorString(ByVal eOperator As SQL.BitwiseOperator) As String

            Select Case eOperator
                Case BitwiseOperator.And
                    Return "&"
                Case BitwiseOperator.Or
                    Return "|"
                Case Else
                    Throw New NotSupportedException(GetType(SQL.BitwiseOperator).Name)
            End Select

        End Function

    End Class

End Namespace

File: SQL/Expressions/SQLBitwiseExpression.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2012. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;

namespace DatabaseObjects.SQL
{
	public enum BitwiseOperator
	{
		@And,
		@Or
	}
		
	public class SQLBitwiseExpression : SQLExpression
	{
		private SQLExpression pobjLeft;
		private BitwiseOperator peOperator;
		private SQLExpression pobjRight;
			
		public SQLBitwiseExpression()
		{
		}
			
		public SQLBitwiseExpression(string strLeftFieldName, BitwiseOperator eOperator, object objRightValue) 
            : this(new SQLFieldExpression(strLeftFieldName), eOperator, new SQLValueExpression(objRightValue))
		{
		}
			
		public SQLBitwiseExpression(string strLeftFieldName, BitwiseOperator eOperator, SQLExpression objRightExpression) 
            : this(new SQLFieldExpression(strLeftFieldName), eOperator, objRightExpression)
		{
		}
			
		public SQLBitwiseExpression(SQLExpression objLeftExpression, BitwiseOperator eOperator, SQLExpression objRightExpression)
		{
			pobjLeft = objLeftExpression;
			peOperator = eOperator;
			pobjRight = objRightExpression;
		}
			
		public SQLExpression LeftExpression
		{
			get
			{
				return pobjLeft;
			}
				
			set
			{
				if (value == null)
					throw new ArgumentNullException();
					
				pobjLeft = value;
			}
		}
			
		public SQLExpression RightExpression
		{
			get
			{
				return pobjRight;
			}
				
			set
			{
				if (value == null)
					throw new ArgumentNullException();
					
				pobjRight = value;
			}
		}
			
		public BitwiseOperator Operator
		{
			get
			{
				return peOperator;
			}
				
			set
			{
				peOperator = value;
			}
		}
			
		internal override string SQL(Serializers.Serializer serializer)
		{
			return serializer.SerializeBitwiseExpression(this);
		}
	}
}
